### DISCUSSION TITLE
add a baffle or stone to the simulation

### DISCUSSION THREAD

**Discussion by QidiMa** [usnistgov/fipy#979] [Q&A]
Hi all,

It should be quite obvious but I'm a beginner with Fipy, apologies! Here is my simulation code, and I want to add a baffle or stone to the simulation .What should I do.

Thanks a lot! Best regards,

```python
###
from fipy import CellVariable, Grid2D, TransientTerm, DiffusionTerm, ExponentialConvectionTerm, Viewer
import fipy as fp
import numpy as np

# create mesh
nx, ny = 100, 100
dx, dy = 1., 1.
mesh = Grid2D(dx=dx, dy=dy, nx=nx, ny=ny)

# velocity
u = np.ones((100, 100))*10
v = np.ones((100, 100))
u_var = CellVariable(mesh=mesh, value=u.flatten())  # 10000
v_var = CellVariable(mesh=mesh, value=v.flatten())  # 10000
u_face_var = u_var.arithmeticFaceValue  # 20200
v_face_var = v_var.arithmeticFaceValue  # 20200
velocity = fp.FaceVariable(mesh=mesh, value=[u_face_var, v_face_var], rank=1)

# define concentration
c = CellVariable(name="Concentration", mesh=mesh, value=0.)

# boundary condition
c.constrain(0., mesh.exteriorFaces)

# parameters
# diffusion coefficient
D = 1

# time setting
dt = 0.1
steps = 50

# source setting
source_duration = 10
source_value = 10.0
r = 4.0
center_x, center_y = nx / 2, ny / 2

# viewer setting
viewer = Viewer(vars=c)

# time iteration
for step in range(steps):
    if step < source_duration:
        source_term = source_value * ((mesh.x - center_x) ** 2 + (mesh.y - center_y) ** 2 <= r ** 2)
    else:
        source_term = 0.

    # define equation
    eq = (TransientTerm(var=c) == DiffusionTerm(coeff=D, var=c) - ExponentialConvectionTerm(coeff=velocity, var=c) + source_term)
    eq.solve(var=c, dt=dt)
    # save picture
    if step % 3 == 0:
        viewer.plot()
    # viewer.plot(filename=f"{step}.png")
```

**Comment by guyer** [comment]
What do you mean by "baffle or stone"? What do you expect it to do, either mathematically (preferable) or physically?

**Reply by QidiMa** [reply]
HELLO!  I want to simulate an area close to the coast, how do I do that?  Thank you.

**Comment by QidiMa** [comment]
Hello, I want to simulate the diffusion of river sewage. If there is obstruction or land ahead, how can I change my simulation? Thank you

**Reply by QidiMa** [reply]
emm.. I guess physics

**Comment by guyer** [comment]
FiPy is a tool for solving partial differential equations on a computer. It is not a tool for telling you what your equations are. That's something you need to bring, using your domain knowledge.

That being said, here are two illustrations that might serve your needs:
- [Code that uses Gmsh to create a mesh with a hole in it](https://github.com/guyer/phasefieldbenchmark-5/blob/master/mesh5b.py) for the simulation of [flow in a channel with an obstruction](https://pages.nist.gov/pfhub/benchmarks/benchmark5-hackathon.ipynb/). This defines the obstruction as being outside the simulation.
- [Discussion on how to prevent diffusion into a region inside the simulation](https://github.com/usnistgov/fipy/discussions/918). This defines the obstruction as being inside the simulation.

**Reply by QidiMa** [reply]
Thank you very much for your reply，
I'll take a few days to mull over and experiment with all the details. 
I'm sorry, maybe my expression was incorrect. I know the equation I want to solve is the two-dimensional convection diffusion equation, but I am not sure how to handle the internal barrier. 
Thank you again.
